### PR TITLE
Feature/add book button fuction promocode

### PR DESCRIPTION
**Pull request by UTALIPOV250618**

### CODE REVIEW

File: src/main/java/kg/peaksoft/ebookm1/entity/book/Book.java [hunk 1]
```diff
@@ -72,4 +74,13 @@ public class Book {
     @OneToOne(cascade = CascadeType.ALL)
     @JoinColumn(name = "favorite_id")
     private Favorite favorite;
+
```
Comment (meerimb) [inline src/main/java/kg/peaksoft/ebookm1/entity/book/Book.java:77]: add @JsonIgnore

File: src/main/java/kg/peaksoft/ebookm1/EbookM1Application.java [hunk 2]
```diff
@@ -3,11 +3,15 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
+
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/EbookM1Application.java:6]: Эмнеге бул класска өзгөртүүлөрдү киригип жатасыздар?

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Operation(summary = "Method all", description = "Allows to get all VENDORS from the database")
+    @GetMapping
+    public List<VendorResponse> getAllVendor() {
+        return userService.getAllUsers();
+    }
+
+    @Operation(summary = "New book added to vendors profile")
+    @PostMapping("/addBook/{id}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:63]: Use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Operation(summary = "Update book of vendors profile")
+    @PutMapping("/updateBook/{vendorId}/{bookId}")
+    public VendorResponse updateBook(@PathVariable long vendorId, @PathVariable long bookId, @RequestBody BookRequest request) {
+        return userService.updateBookVendor(vendorId, bookId, request);
+    }
+
+    @Operation(summary = "Delete book from vendors profile")
+    @DeleteMapping("/deleteBook/{vendorId}/{bookId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:75]: Use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Operation(summary = "New promocode added to vendors profile")
+    @PostMapping("/addPromocode/{id}")
+    public VendorResponse addPromocode(@PathVariable long id, @RequestBody PromocodeRequest promocodeRequest) {
+        return userService.addPromocode(promocodeRequest, id);
+    }
+
+    @Operation(summary = "Update promocode of vendors profile")
+    @PutMapping("/updatePromocode/{vendorId}/{promoCodeId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:87]: Use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/dto/promocode/PromocodeRequest.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDate;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/dto/promocode/PromocodeRequest.java:12]: Use @Getter and @Setter instead of @Data

File: src/main/java/kg/peaksoft/ebookm1/dto/promocode/PromocodeResponse.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDate;
+
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/dto/promocode/PromocodeResponse.java:12]: Use @Getter and @Setter instead of @Data

File: src/main/java/kg/peaksoft/ebookm1/entity/Promocode.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import kg.peaksoft.ebookm1.entity.book.Book;
+import lombok.Data;
+
+import javax.persistence.*;
+import java.time.LocalDate;
+import java.util.List;
+
+@Entity
+@Table(name = "promocodes")
+@Data
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/entity/Promocode.java:14]: Use @Getter and @Setter instead of @Data

File: src/main/resources/application-dev.properties [hunk 9]
```diff
@@ -1,8 +1,7 @@
-spring.datasource.url=jdbc:postgresql://localhost:5432/ebook
+spring.datasource.url=jdbc:postgresql://localhost:5432/eBook1
 spring.jpa.database-platform=org.hibernate.dialect.PostgreSQL10Dialect
 spring.datasource.username=postgres
-spring.datasource.password=bekbolot
-spring.jpa.hibernate.ddl-auto=update
+spring.datasource.password=google224
+spring.jpa.hibernate.ddl-auto=create-drop
 spring.jpa.show-sql=true
```
Comment (ChyngyzIshen) [inline src/main/resources/application-dev.properties:7]: Don't add application-dev.properties file to the pull request

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -1,8 +1,8 @@
-spring.datasource.url=jdbc:postgresql://localhost:5432/ebook
+spring.datasource.url=jdbc:postgresql://localhost:5432/eBook1
 spring.datasource.username=postgres
 spring.datasource.password=postgres
 spring.jpa.hibernate.ddl-auto=create
-spring.jpa.show-sql=false
+spring.jpa.show-sql=true
 spring.jpa.database-platform=org.hibernate.dialect.PostgreSQL10Dialect
```
Comment (ChyngyzIshen) [inline src/main/resources/application.properties:7]: Don't touch application.properties file at all, except if you will not add new configs

**Review by ChyngyzIshen — COMMENTED** [review]
Comments

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 11]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
-    @Operation(summary = "<Method delete by id", description = "User with role ADMIN and VENDOR can deleted")
+    @Operation(summary = "Method delete by id", description = "User with role ADMIN and VENDOR can deleted")
     @PreAuthorize("hasAnyAuthority('ADMIN', 'VENDOR')")
     @DeleteMapping("{id}")
     public VendorResponse deleteById(@PathVariable Long id) {
         return service.deleteById(id);
     }
+    @Operation(summary = "Method to add new book", description = "Vendor can add new book to his profile")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PostMapping("/{id}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:39]: @PostMapping("/{id}") same method

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 12]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
+    @Operation(summary = "Method to delete book", description = "Vendor and ADMIN can delete book from his book list")
+    @PreAuthorize("hasAnyAuthority('ADMIN', 'VENDOR')")
+    @DeleteMapping("/{vendorId}/{bookId}")
+    public VendorResponse deleteBook(@PathVariable long vendorId, @PathVariable long bookId) {
+        return service.deleteBookVendor(vendorId, bookId);
+    }
+
+    @Operation(summary = "Method to add new promocode", description = "Vendor can add new promocode to his profile")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PostMapping("/{id}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:60]: @PostMapping("/{id}") same method will not work

File: src/main/java/kg/peaksoft/ebookm1/repository/BookRepository.java [hunk 13]
```diff
@@ -3,7 +3,7 @@
 import kg.peaksoft.ebookm1.entity.book.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
-import org.springframework.stereotype.Service;
+
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/repository/BookRepository.java:6]: Unnecessary whitespace

File: src/main/java/kg/peaksoft/ebookm1/repository/PromocodeRepository.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package kg.peaksoft.ebookm1.repository;
+
+import kg.peaksoft.ebookm1.entity.Promocode;
+import org.springframework.data.jpa.repository.JpaRepository;
+
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/repository/PromocodeRepository.java:5]: @Repository?

File: src/main/java/kg/peaksoft/ebookm1/service/PromoService.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
+package kg.peaksoft.ebookm1.service;
+
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class PromoService {
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/service/PromoService.java:9]: Why empty class?

**Review by ChyngyzIshen — COMMENTED** [review]
comments

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 16]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
+    @Operation(summary = "Method to update vendor's promocode", description = "Vendor can update his books from his promocode list")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PutMapping("/update-promo/{vendorId}/{promoCodeId}")
+    public VendorResponse updatePromoCode(@PathVariable long vendorId, @PathVariable long promoCodeId, @RequestBody PromocodeRequest promocodeRequest) {
+        return service.updatePromocode(promocodeRequest, vendorId, promoCodeId);
+    }
+
+    @Operation(summary = "Method to delete book", description = "Vendor and ADMIN can delete promocode from his book list")
+    @PreAuthorize("hasAnyAuthority('ADMIN', 'VENDOR')")
+    @DeleteMapping("/delete-promo/{vendorId}/{promoCodeId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:74]: please use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 17]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
+    @Operation(summary = "Method to add new promocode", description = "Vendor can add new promocode to his profile")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PostMapping("/add-promo/{id}")
+    public VendorResponse addPromoCode(@PathVariable long id, @RequestBody PromocodeRequest promocodeRequest) {
+        return service.addPromocode(promocodeRequest, id);
+    }
+
+    @Operation(summary = "Method to update vendor's promocode", description = "Vendor can update his books from his promocode list")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PutMapping("/update-promo/{vendorId}/{promoCodeId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:67]: please use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 18]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
+    @Operation(summary = "Method to delete book", description = "Vendor and ADMIN can delete book from his book list")
+    @PreAuthorize("hasAnyAuthority('ADMIN', 'VENDOR')")
+    @DeleteMapping("/delete-book/{vendorId}/{bookId}")
+    public VendorResponse deleteBook(@PathVariable long vendorId, @PathVariable long bookId) {
+        return service.deleteBookVendor(vendorId, bookId);
+    }
+
+    @Operation(summary = "Method to add new promocode", description = "Vendor can add new promocode to his profile")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PostMapping("/add-promo/{id}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:60]: please use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 19]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
+    @Operation(summary = "Method to update vendor's book", description = "Vendor can update his books from his book list")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PutMapping("/update-book/{vendorId}/{bookId}")
+    public VendorResponse updateBook(@PathVariable long vendorId, @PathVariable long bookId, @RequestBody BookRequest request) {
+        return service.updateBookVendor(vendorId, bookId, request);
+    }
+
+    @Operation(summary = "Method to delete book", description = "Vendor and ADMIN can delete book from his book list")
+    @PreAuthorize("hasAnyAuthority('ADMIN', 'VENDOR')")
+    @DeleteMapping("/delete-book/{vendorId}/{bookId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:53]: please use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 20]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
+    @Operation(summary = "Method to add new book", description = "Vendor can add new book to his profile")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PostMapping("/add-new-book/{id}")
+    public VendorResponse addBookToVendor(@PathVariable long id, @RequestBody BookRequest request) {
+        return service.addBookToVendor(request, id);
+    }
+
+    @Operation(summary = "Method to update vendor's book", description = "Vendor can update his books from his book list")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PutMapping("/update-book/{vendorId}/{bookId}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:46]: please use proper endpoint naming convention

File: src/main/java/kg/peaksoft/ebookm1/api/VendorController.java [hunk 21]
```diff
@@ -26,10 +28,52 @@ public VendorResponse updateById(@PathVariable Long id, @RequestBody VendorReque
...
-    @Operation(summary = "<Method delete by id", description = "User with role ADMIN and VENDOR can deleted")
+    @Operation(summary = "Method delete by id", description = "User with role ADMIN and VENDOR can deleted")
     @PreAuthorize("hasAnyAuthority('ADMIN', 'VENDOR')")
     @DeleteMapping("{id}")
     public VendorResponse deleteById(@PathVariable Long id) {
         return service.deleteById(id);
     }
+    @Operation(summary = "Method to add new book", description = "Vendor can add new book to his profile")
+    @PreAuthorize("hasAnyAuthority('VENDOR')")
+    @PostMapping("/add-new-book/{id}")
```
Comment (ChyngyzIshen) [inline src/main/java/kg/peaksoft/ebookm1/api/VendorController.java:39]: please use proper endpoint naming convention

**Review by ChyngyzIshen — COMMENTED** [review]
Comments still not resolved
